release-automations/create_release_branch.py

```diff
@@ -0,0 +1,527 @@
+    python create_release_branch.py --release_branch <branch> [options]
+
+Options:
+    --release_branch         Name of the branch to create (required).
+    --apitoken              API token for authentication with upstream repositories.
+    --staging               Use staging environment.
+    --prod                  Use production environment.
+    --mailing_list          List of email addresses to notify.
+
+Note:
+    - This script assumes a specific repository structure and the presence of .gitmodules.
+"""
+import argparse
+import configparser
+import glob
+import logging
+import os
+import shutil
+import subprocess
+import tempfile
+import pandas as pd
+import smtplib
+from email.mime.text import MIMEText
+from email.mime.multipart import MIMEMultipart
+
+logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)s %(message)s')
+
+class rockbrachingAutomation():
+
+    def __init__(self, args):
+        
+        self.release_branch = args.release_branch
+        if args.apitoken:
+            self.apitoken = args.apitoken
+        else:
+            self.apitoken = None
+        self.is_staging = bool(args.staging)
+        self.is_prod = bool(args.prod)
+        
+        if args.staging:
+            self.gh_hostname = "github.amd.com"
+            self.org = "ROCm-Staging"
+        if args.prod:
+            self.gh_hostname = "github.com"
+            self.org = "ROCm"
+        # Derive environment label for notifications
+        if self.is_staging:
+            self.environment = "Staging"
+        elif self.is_prod:
+            self.environment = "Prod"
+        else:
+            self.environment = "Unknown"
+        
+        if args.prod and args.staging:
+            logging.error("You cannot pass both --staging(-S) and --prod(-P); choose one")
+            exit(1)
+        if args.mailing_list:
+            self.mailing_list = args.mailing_list
+        else:
+            self.mailing_list = None
+
+
+    # Helper function to run shell commands via subprocess.run.
+    # Return exit_code and output (if capture_output is True).
+    def run(self, cmd, cwd=None, capture_output=False, check=True):
+        result = subprocess.run(
+            cmd,
+            shell=True,
+            cwd=cwd,
+            capture_output=capture_output,
+            text=True,
+        )
+        if capture_output:
+            if check and result.returncode != 0:
+                raise subprocess.CalledProcessError(result.returncode, cmd)
+            return (result.returncode, result.stdout.strip())
+        else:
+            if check and result.returncode != 0:
+                raise subprocess.CalledProcessError(result.returncode, cmd)
+            return (result.returncode, None)
+        
+
+
+    # Check out source code and returns a path to where theRock is checked out, projects dict, and
+    def checkout_source(self):
+        the_rock_url = f"https://{self.apitoken}@{self.gh_hostname}/{self.org}/TheRock.git"
```

Comment:
If we drop the staging/prod code, I think we can just use standard SSH / HTTPS cloning for repositories, without needing an API token.